File: src/com/l2kt/gameserver/model/itemcontainer/PetInventory.java
package com.l2kt.gameserver.model.itemcontainer;

import com.l2kt.commons.random.Rnd;
import com.l2kt.gameserver.model.actor.instance.Pet;
import com.l2kt.gameserver.model.actor.instance.Player;
import com.l2kt.gameserver.model.item.instance.ItemInstance;
import com.l2kt.gameserver.model.item.type.EtcItemType;

public class PetInventory extends Inventory
{
	private final Pet _owner;
	
	public PetInventory(Pet owner)
	{
		_owner = owner;
	}
	
	@Override
	public Pet getOwner()
	{
		return _owner;
	}
	
	@Override
	public int getOwnerId()
	{
		int id;
		try
		{
			id = _owner.getOwner().getObjectId();
		}
		catch (NullPointerException e)
		{
			return 0;
		}
		return id;
	}
	
	@Override
	protected void refreshWeight()
	{
		super.refreshWeight();
		
		getOwner().updateAndBroadcastStatus(1);
		getOwner().sendPetInfosToOwner();
	}
	
	public boolean validateCapacity(ItemInstance item)
	{
		int slots = 0;
		
		if (!(item.isStackable() && getItemByItemId(item.getItemId()) != null) && item.getItemType() != EtcItemType.HERB)
			slots++;
		
		return validateCapacity(slots);
	}
	
	@Override
	public boolean validateCapacity(int slots)
	{
		return _items.size() + slots <= _owner.getInventoryLimit();
	}
	
	public boolean validateWeight(ItemInstance item, int count)
	{
		return validateWeight(count * item.getItem().getWeight());
	}
	
	@Override
	public boolean validateWeight(int weight)
	{
		return _totalWeight + weight <= _owner.getMaxLoad();
	}
	
	@Override
	protected ItemInstance.ItemLocation getBaseLocation()
	{
		return ItemInstance.ItemLocation.PET;
	}
	
	@Override
	protected ItemInstance.ItemLocation getEquipLocation()
	{
		return ItemInstance.ItemLocation.PET_EQUIP;
	}
	
	@Override
	public void deleteMe()
	{
		final Player petOwner = getOwner().getOwner();
		if (petOwner != null)
		{
			for (ItemInstance item : _items)
			{
				if (petOwner.getInventory().validateCapacity(1))
					getOwner().transferItem("return", item.getObjectId(), item.getCount(), petOwner.getInventory(), petOwner, getOwner());
				else
				{
					final ItemInstance droppedItem = dropItem("drop", item.getObjectId(), item.getCount(), petOwner, getOwner());
					droppedItem.dropMe(getOwner(), getOwner().getX() + Rnd.INSTANCE.get(-70, 70), getOwner().getY() + Rnd.INSTANCE.get(-70, 70), getOwner().getZ() + 30);
				}
				
			}
		}
		_items.clear();
	}
}

File: src/com/l2kt/gameserver/model/itemcontainer/PetInventory.kt
package com.l2kt.gameserver.model.itemcontainer

import com.l2kt.commons.random.Rnd
import com.l2kt.gameserver.model.actor.instance.Pet
import com.l2kt.gameserver.model.item.instance.ItemInstance
import com.l2kt.gameserver.model.item.type.EtcItemType

class PetInventory(public override val owner: Pet) : Inventory() {

    override val ownerId: Int
        get() {
            return owner.owner?.objectId ?: 0
        }

    override val baseLocation: ItemInstance.ItemLocation
        get() = ItemInstance.ItemLocation.PET

    override val equipLocation: ItemInstance.ItemLocation
        get() = ItemInstance.ItemLocation.PET_EQUIP

    override fun refreshWeight() {
        super.refreshWeight()

        owner.updateAndBroadcastStatus(1)
        owner.sendPetInfosToOwner()
    }

    fun validateCapacity(item: ItemInstance): Boolean {
        var slots = 0

        if (!(item.isStackable && getItemByItemId(item.itemId) != null) && item.itemType !== EtcItemType.HERB)
            slots++

        return validateCapacity(slots)
    }

    override fun validateCapacity(slots: Int): Boolean {
        return _items.size + slots <= owner.inventoryLimit
    }

    fun validateWeight(item: ItemInstance, count: Int): Boolean {
        return validateWeight(count * item.item.weight)
    }

    override fun validateWeight(weight: Int): Boolean {
        return totalWeight + weight <= owner.maxLoad
    }

    override fun deleteMe() {
        val petOwner = owner.owner
        if (petOwner != null) {
            for (item in _items) {
                if (petOwner.inventory!!.validateCapacity(1))
                    owner.transferItem("return", item.objectId, item.count, petOwner.inventory, petOwner, owner)
                else {
                    val droppedItem = dropItem("drop", item.objectId, item.count, petOwner, owner)
                    droppedItem?.dropMe(owner, owner.x + Rnd[-70, 70], owner.y + Rnd[-70, 70], owner.z + 30)
                }

            }
        }
        _items.clear()
    }
}